include/cetl/pf17/variant.hpp

```diff
@@ -0,0 +1,1270 @@
+{
+    // Instead of generating a multidimensional vtable as it is commonly done, we use recursive visiting;
+    // this allows us to achieve a similar result with much less code at the expense of one extra call indirection
+    // per variant. The recursive structure below builds a Cartesian product of the variant combinations
+    // one level at a time starting with the leftmost variant. The complexity is one table jump per variant level.
+    return visit(
+        [&](auto&& hh) {  // https://twitter.com/PavelKirienko/status/1761525562370040002
+            return visit(
+                [&](auto&&... tt) {
+                    return std::forward<F>(fun)(std::forward<decltype(hh)>(hh), std::forward<decltype(tt)>(tt)...);
+                },
+                std::forward<Vs>(vars)...);
+        },
+        std::forward<V0>(var0));
+}
+
+}  // namespace var
+}  // namespace detail
+
+// --------------------------------------------------------------------------------------------------------------------
+
+/// Implementation of \ref std::variant_alternative.
+/// This implementation also accepts other typelist-parameterized classes, such as \ref std::variant.
+template <size_t, typename>
+struct variant_alternative;
+template <size_t N, template <typename...> class V, typename... Ts>
+struct variant_alternative<N, V<Ts...>>
+{
+    static_assert(N < sizeof...(Ts), "Variant type index out of range");
+    using type = detail::var::nth_type<N, Ts...>;
+};
+template <size_t N, template <typename...> class V, typename... Ts>
+struct variant_alternative<N, const V<Ts...>>
+{
+    using type = const typename variant_alternative<N, V<Ts...>>::type;
+};
+
+/// Implementation of \ref std::variant_alternative_t.
+/// This implementation also accepts other typelist-parameterized classes, such as \ref std::variant.
+template <size_t N, typename V>
+using variant_alternative_t = typename variant_alternative<N, V>::type;
+
+// --------------------------------------------------------------------------------------------------------------------
+
+/// Implementation of \ref std::variant_size.
+/// This implementation also accepts other typelist-parameterized classes, such as \ref std::variant.
+template <typename>
+struct variant_size;
+template <typename V>
+struct variant_size<const V> : variant_size<V>
+{};
+template <template <typename...> class V, typename... Ts>
+struct variant_size<V<Ts...>> : std::integral_constant<std::size_t, sizeof...(Ts)>
+{};
+
+/// Implementation of \ref std::variant_size_v.
+/// This implementation also accepts other typelist-parameterized classes, such as \ref std::variant.
+template <typename V>
+constexpr size_t variant_size_v = variant_size<V>::value;
+
+// --------------------------------------------------------------------------------------------------------------------
+
+/// An implementation of \ref std::variant.
+///
+/// The valueless state cannot occur unless exceptions are enabled.
```

Comment:
Missing docs for @param T

Reply:
I added it but note that the docs are minimal because this is just a standard variant